### PR TITLE
Highlighting in the doc that square root comes before adding epsilon

### PR DESCRIPTION
**Pull request by vincentqb**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26735 Highlighting in the doc that square root comes before adding epsilon**

Differential Revision: [D17558505](https://our.internmc.facebook.com/intern/diff/D17558505)

### CODE REVIEW

File: torch/optim/rmsprop.py [hunk 1]
```diff
@@ -11,6 +11,9 @@ class RMSprop(Optimizer):
     The centered version first appears in `Generating Sequences
     With Recurrent Neural Networks <https://arxiv.org/pdf/1308.0850v5.pdf>`_.
 
+    Note that the implementation here takes the square root of the gradient average before
+    adding epsilon.
```
Comment (zou3519) [inline torch/optim/rmsprop.py]: nits: It would be nice to have an equation like one of those that you showed me offline; equations are easier to read sometimes. Also, it might be good to document that this differs from the TF implementation specifically because I'm not sure if someone using RMSprop knows exactly where TF does the square root without investigation.

**Review by zou3519 — APPROVED** [review]
lgtm.

File: torch/optim/rmsprop.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/rmsprop.py]: Good points. I've added a note about tensorflow and the effective learning rate for RMSProp as `\alpha/(\sqrt{v} + \epsilon)`.

File: torch/optim/rmsprop.py [hunk 2]
```diff
@@ -3,14 +3,20 @@
...
+    r"""Implements RMSprop algorithm.
 
     Proposed by G. Hinton in his
     `course <http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf>`_.
 
     The centered version first appears in `Generating Sequences
     With Recurrent Neural Networks <https://arxiv.org/pdf/1308.0850v5.pdf>`_.
 
+    The implementation here takes the square root of the gradient average before
+    adding epsilon (note that Tensor Flow interchange these two operations). The effective
```
Comment (zou3519) [inline torch/optim/rmsprop.py]: TensorFlow is one word, I think. Also, I think this should be "note that TensorFlow interchanges these two operations"

Comment (vincentqb) [reply torch/optim/rmsprop.py]: Oops! Thanks :)

**Comment by bonlime** [comment]
@vincentqb @zou3519 
Hi, 
May I ask is there any particular reason the current implementation of RMSprop adds epsilon after sqrt? 
From my experience training with default PyTorch RMSprop + FP16 (using apex) tends to produce `nan` during backprop. I'm sure @rwightman also experienced such issues. Training EfficientNets, for example, is impossible. 

By moving epsilon inside sqrt everything works like a charm. There are also mathematical proofs that adding eps after sqrt is not enough to prevent overflow. [Check this](https://arxiv.org/pdf/1911.05920.pdf) paper at p.6 bottom right. The topic of the paper is slightly different but their conclusion is also applicable here.

**Comment by bonlime** [comment]
According to https://github.com/pytorch/pytorch/issues/23796 other people also had problems with RMSprop behavior in PyTorch. As far as I see you want the current implementation to be consistent with Adam / AdamW but they also tend to produce `nan`'s! 
It's a very tricky issue and it's hard to show that proposed implementation is superior to the current one by tests. But I know a bunch of people who successfully used adjusted versions of Adam / RMSprop to avoid nans
